src/Block/ProfileMenuBlockService.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\UserBundle\Block;
+
+use Knp\Menu\ItemInterface;
+use Knp\Menu\Provider\MenuProviderInterface;
+use Sonata\BlockBundle\Block\BlockContextInterface;
+use Sonata\BlockBundle\Block\Service\MenuBlockService;
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @author Hugo Briand <briand@ekino.com>
+ */
+final class ProfileMenuBlockService extends MenuBlockService
+{
+    private $menuBuilder;
+
+    /**
+     * @param object $menuBuilder
+     */
+    public function __construct(string $name, EngineInterface $templating, MenuProviderInterface $menuProvider, $menuBuilder)
+    {
+        parent::__construct($name, $templating, $menuProvider, []);
+
+        if (!\is_object($menuBuilder) || !\is_callable([$menuBuilder, 'createProfileMenu'])) {
+            throw new \InvalidArgumentException(
+                'Argument 4 should be object with public function "createProfileMenu(array $itemOptions = [])"'
+            );
+        }
+
+        $this->menuBuilder = $menuBuilder;
+    }
+
+    public function getName(): string
+    {
+        return 'Ecommerce Profile Menu';
+    }
+
+    public function configureSettings(OptionsResolver $resolver): void
```

Comment:
Even if the PHP version allows to change the signatures defined at parent, we should avoid to do that.

```suggestion
    public function configureSettings(OptionsResolver $resolver)
```

Reply:
https://github.com/sonata-project/SonataUserBundle/pull/1136#discussion_r363974369 - little fonfusing, but you are right.

Reply:
:thinking: :thinking: :thinking: why should we avoid doing that? In case the parent later adds a type hint that is not that one?

Reply:
For me we should update return type hints in final classes if parent return void or only one posibility. `MenuBlockService:getMenu` return `ItemInterface|string`.

We should avoid return type hints in classess where parents haven't it and user can extends them.

Reply:
>    why should we avoid doing that? In case the parent later adds a type hint that is not that one?

Yes, and to be consistent with the behavior of the argument declarations (https://3v4l.org/9Kc60).

Reply:
So we can add:
```php 
public function configureSettings(OptionsResolver $resolver) 
public function configureSettings(OptionsResolver $resolver): void 
```
but we cant add:
```php 
public function __construct($name, EngineInterface $templating, ... )
public function __construct(string $name, EngineInterface $templating, ... )
```

Reply:
`__construct()` method is a special case, it allows to change their arguments and their types.